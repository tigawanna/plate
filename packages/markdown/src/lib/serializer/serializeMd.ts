import type { Descendant, SlateEditor } from 'platejs';

import remarkStringify from 'remark-stringify';
import { type Plugin, unified } from 'unified';

import type { AllowNodeConfig } from '../MarkdownPlugin';
import type { MdRoot } from '../mdast';
import type { MdRules, PlateType } from '../types';

import { convertNodesSerialize } from './convertNodesSerialize';
import { getMergedOptionsSerialize } from './utils/getMergedOptionsSerialize';
export type SerializeMdOptions = {
  allowedNodes?: PlateType[] | null;
  allowNode?: AllowNodeConfig;
  disallowedNodes?: PlateType[] | null;
  editor?: SlateEditor;
  preserveEmptyParagraphs?: boolean;
  remarkPlugins?: Plugin[];
  rules?: MdRules;
  spread?: boolean;
  value?: Descendant[];
  withBlockId?: boolean;
};

/** Serialize the editor value to Markdown. */
export const serializeMd = (
  editor: SlateEditor,
  options?: Omit<SerializeMdOptions, 'editor'>
) => {
  const mergedOptions = getMergedOptionsSerialize(editor, options);

  const { remarkPlugins, value } = mergedOptions;

  const toRemarkProcessor = unified()
    .use(remarkPlugins ?? [])
    .use(remarkStringify, {
      emphasis: '_',
    });

  const mdast = slateToMdast({
    children: value!,
    options: mergedOptions,
  });

  return toRemarkProcessor.stringify(mdast);
};

const slateToMdast = ({
  children,
  options,
}: {
  children: Descendant[];
  options: SerializeMdOptions;
}): MdRoot => {
  const processedChildren = convertNodesSerialize(
    children,
    options,
    true // isBlock = true for top-level elements
  ) as MdRoot['children'];

  const ast = {
    children: processedChildren,
    type: 'root',
  } as MdRoot;
  return ast;
};
