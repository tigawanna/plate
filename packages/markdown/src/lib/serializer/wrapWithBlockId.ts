import type { unistLib } from '../types';

/**
 * Wraps an mdast node with a block element containing an ID attribute. Used for
 * preserving block IDs when serializing to markdown.
 *
 * @param mdastNode - The mdast node to wrap
 * @param nodeId - The ID to attach to the block element
 * @returns The wrapped mdast node with block element and ID attribute
 */
export const wrapWithBlockId = (
  mdastNode: unistLib.Node,
  nodeId: string
): unistLib.Node => {
  return {
    attributes: [
      {
        name: 'id',
        type: 'mdxJsxAttribute',
        value: String(nodeId),
      },
    ],
    children: [mdastNode],
    data: {
      _mdxExplicitJsx: true,
    },
    name: 'block',
    type: 'mdxJsxFlowElement',
  } as any;
};
