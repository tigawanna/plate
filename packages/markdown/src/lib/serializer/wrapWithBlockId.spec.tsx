/** @jsx jsx */

import { jsx } from '@platejs/test-utils';

import { createTestEditor } from '../__tests__/createTestEditor';
import { serializeMd } from './serializeMd';
import { wrapWithBlockId } from './wrapWithBlockId';

jsx;

const editor = createTestEditor();

describe('wrapWithBlockId', () => {
  describe('unit tests', () => {
    it('should wrap a node with block element and ID attribute', () => {
      const mdastNode = {
        children: [{ type: 'text', value: 'Hello world' }],
        type: 'paragraph',
      };

      const result = wrapWithBlockId(mdastNode, 'test-id');

      expect(result).toEqual({
        attributes: [
          {
            name: 'id',
            type: 'mdxJsxAttribute',
            value: 'test-id',
          },
        ],
        children: [mdastNode],
        data: {
          _mdxExplicitJsx: true,
        },
        name: 'block',
        type: 'mdxJsxFlowElement',
      });
    });

    it('should convert numeric IDs to strings', () => {
      const mdastNode = {
        children: [{ type: 'text', value: 'Test' }],
        type: 'paragraph',
      };

      const result = wrapWithBlockId(mdastNode, '123') as any;

      expect(result.attributes[0].value).toBe('123');
    });
  });

  describe('integration with serializeMd', () => {
    it('should wrap nodes with IDs in block elements when withBlockId is true', () => {
      const slateNodes = [
        {
          id: '123',
          children: [{ text: 'Hello world' }],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, {
        value: slateNodes,
        withBlockId: true,
      });

      expect(result).toMatchSnapshot();
    });

    it('should not wrap nodes without IDs', () => {
      const slateNodes = [
        {
          children: [{ text: 'No ID here' }],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, {
        value: slateNodes,
        withBlockId: true,
      });

      expect(result).toMatchSnapshot();
    });

    it('should handle complex nested structures', () => {
      const slateNodes = [
        {
          id: 'para-with-marks',
          children: [
            { text: 'Text with ' },
            { bold: true, text: 'bold' },
            { text: ' and ' },
            { italic: true, text: 'italic' },
            { text: ' marks' },
          ],
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, {
        value: slateNodes,
        withBlockId: true,
      });

      expect(result).toMatchSnapshot();
    });

    it('should not wrap nested block elements in tables', () => {
      const slateNodes = (
        <htable id="wX11A61CZN">
          <htr id="Q1LgIIFN3Q">
            <htd id="9gnGaBxtE_">
              <hp id="U8i7WkT4O4">td1</hp>
            </htd>
            <htd id="gPanekRFP8">
              <hp id="t8-eEle3Y-">td2</hp>
            </htd>
          </htr>
          <htr id="xjI2n84Qz9">
            <htd id="zsCvPVVyaI">
              <hp id="SDNQL3h_Or">td3</hp>
            </htd>
            <htd id="KZnBaSppxO">
              <hp id="NuCtsYJdZP">td4</hp>
            </htd>
          </htr>
        </htable>
      );

      const result = serializeMd(editor as any, {
        value: [slateNodes],
        withBlockId: true,
      });

      expect(result).toMatch(
        `<block id="wX11A61CZN">
  | td1 | td2 |
  | --- | --- |
  | td3 | td4 |
</block>`
      );
    });

    it('should wrap indent lists with IDs in block elements when withBlockId is true', () => {
      const slateNodes = [
        {
          id: 'list-item-1',
          children: [{ text: 'Item 1' }],
          indent: 1,
          listStyleType: 'disc',
          type: 'p',
        },
        {
          id: 'list-item-2',
          children: [{ text: 'Item 2' }],
          indent: 1,
          listStyleType: 'disc',
          type: 'p',
        },
        {
          id: 'list-item-3',
          children: [{ text: 'Nested item' }],
          indent: 2,
          listStyleType: 'disc',
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, {
        value: slateNodes,
        withBlockId: true,
      });

      expect(result).toMatchSnapshot();
    });

    it('should wrap ordered indent lists with IDs when withBlockId is true', () => {
      const slateNodes = [
        {
          id: 'ordered-1',
          children: [{ text: 'First' }],
          indent: 1,
          listStyleType: 'decimal',
          type: 'p',
        },
        {
          id: 'ordered-2',
          children: [{ text: 'Second' }],
          indent: 1,
          listStyleType: 'decimal',
          type: 'p',
        },
      ];

      const result = serializeMd(editor as any, {
        value: slateNodes,
        withBlockId: true,
      });

      expect(result).toMatchSnapshot();
    });
  });
});
