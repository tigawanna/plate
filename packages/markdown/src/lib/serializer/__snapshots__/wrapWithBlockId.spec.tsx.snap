// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapWithBlockId integration with serializeMd should handle complex nested structures 1`] = `
"<block id="para-with-marks">
  Text with **bold** and _italic_ marks
</block>
"
`;

exports[`wrapWithBlockId integration with serializeMd should not wrap nodes without IDs 1`] = `
"No ID here
"
`;

exports[`wrapWithBlockId integration with serializeMd should wrap indent lists with IDs in block elements when withBlockId is true 1`] = `
"<block id="list-item-1">
  * Item 1
</block>

<block id="list-item-2">
  * Item 2
</block>

<block id="list-item-3">
  * Nested item
</block>
"
`;

exports[`wrapWithBlockId integration with serializeMd should wrap nodes with IDs in block elements when withBlockId is true 1`] = `
"<block id="123">
  Hello world
</block>
"
`;

exports[`wrapWithBlockId integration with serializeMd should wrap ordered indent lists with IDs when withBlockId is true 1`] = `
"<block id="ordered-1">
  1. First
</block>

<block id="ordered-2">
  2. Second
</block>
"
`;
